File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/Playbook.kt
package com.appautopsy.analysis.core

import com.appautopsy.analysis.model.PatternMatch
import com.appautopsy.analysis.model.PlaybookStep
import com.appautopsy.analysis.model.Verdict
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Action playbook (F15): ordered, actionable steps for the verdict.
 *
 * A verdict doesn't change behaviour; an instruction does. Keyed by band and by
 * which patterns matched — `red` picks the most specific playbook available.
 *
 * Contributes 0 points and never alters the verdict. Purely presentational.
 *
 * Steps arrive as `step_key` plus params; the text itself is resolved by the
 * Catalog at render time, so changing language changes the playbook for free.
 * The playbook definitions live in the English file only — the step *keys* are
 * language-neutral, and each language ships its own `strings` block.
 */
object Playbook {

    /** Which playbook list to use, given the verdict and what fired. */
    fun keyFor(
        verdict: Verdict,
        patterns: List<PatternMatch>,
        repackagingDetected: Boolean,
    ): String {
        if (verdict != Verdict.RED) return "${verdict.id}.default"

        val ids = patterns.map { it.id }.toSet()
        return when {
            "otp_stealer" in ids -> "red.otp_stealer"
            "banking_trojan" in ids -> "red.banking_trojan"
            repackagingDetected || "repackaged_signer" in ids -> "red.repackaged_signer"
            else -> "red.default"
        }
    }

    /**
     * Builds the steps for a result. Falls back to `red.default` when a key has
     * no list, because an empty playbook for a dangerous app is worse than a
     * generic one.
     */
    fun steps(
        definitions: Map<String, List<StepDef>>,
        verdict: Verdict,
        patterns: List<PatternMatch>,
        appLabel: String?,
        repackagingDetected: Boolean,
    ): List<PlaybookStep> {
        val key = keyFor(verdict, patterns, repackagingDetected)
        val raw = definitions[key] ?: definitions["red.default"] ?: return emptyList()
        val label = appLabel?.takeIf { it.isNotBlank() } ?: "this app"

        return raw.map { def ->
            val params = if ("app" in def.params) mapOf("app" to label) else emptyMap()
            PlaybookStep(stepKey = def.stepKey, params = params)
        }
    }

    data class StepDef(val stepKey: String, val params: List<String>)

    /**
     * Parses the `playbooks` block of playbook_en.json. Loaded once at startup
     * alongside the Catalog; a scan never re-reads it.
     */
    fun loadDefinitions(text: String): Map<String, List<StepDef>> {
        val json = Json { ignoreUnknownKeys = true }
        val playbooks = json.parseToJsonElement(text).jsonObject["playbooks"]!!.jsonObject

        return playbooks.mapValues { (_, stepsElement) ->
            stepsElement.jsonArray.map { stepElement ->
                val step = stepElement.jsonObject
                StepDef(
                    stepKey = step["step_key"]!!.jsonPrimitive.content,
                    params = step["params"]?.jsonArray
                        ?.map { it.jsonPrimitive.content }
                        ?: emptyList(),
                )
            }
        }
    }
}

File: backend/app/core/playbook.py
"""Action playbook generator (F15).

Generates ordered, actionable steps based on verdict band and matched patterns.
Contributes 0 points and never alters the verdict.
"""

from __future__ import annotations

import json
from functools import lru_cache

from app.config import DATA_DIR
from app.models.schemas import PatternMatch, PlaybookStep, Verdict


@lru_cache(maxsize=1)
def _load_playbooks() -> dict:
    with (DATA_DIR / "playbook_en.json").open(encoding="utf-8") as f:
        return json.load(f).get("playbooks", {})


def get_playbook(
    verdict: Verdict,
    patterns: list[PatternMatch],
    app_label: str = "this app",
    repackaging_detected: bool = False,
) -> list[PlaybookStep]:
    playbooks = _load_playbooks()

    # Determine playbook key
    key = f"{verdict}.default"
    if verdict == "red":
        matched_ids = {p.id for p in patterns}
        if "otp_stealer" in matched_ids:
            key = "red.otp_stealer"
        elif "banking_trojan" in matched_ids:
            key = "red.banking_trojan"
        elif repackaging_detected or "repackaged_signer" in matched_ids:
            key = "red.repackaged_signer"

    raw_steps = playbooks.get(key, playbooks.get("red.default", []))
    steps: list[PlaybookStep] = []

    for item in raw_steps:
        step_key = item["step_key"]
        params_needed = item.get("params", [])
        params_dict = {}
        if "app" in params_needed:
            params_dict["app"] = app_label or "this app"

        steps.append(PlaybookStep(step_key=step_key, params=params_dict))

    return steps
